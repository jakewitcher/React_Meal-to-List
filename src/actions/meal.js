import uuid from 'uuid';

export const addMeal = (
    {
        name = '',
        itemList = [],
    } = {}
) => ({
    type: 'ADD_MEAL',
    meal: {
        id: uuid(),
        name,
        itemList,
    }
});

export const deleteMeal = (id = '') => ({
    type: 'DELETE_MEAL',
    id,
});

export const editMeal = (
    {
        name = '',
        id = '',
        itemList = [],
    } = {}
) => ({
    type: 'EDIT_MEAL',
    meal: {
        id,
        name,
        itemList,
    }
});