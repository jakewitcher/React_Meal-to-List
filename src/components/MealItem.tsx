import * as React from "react";
import { Item } from "../store/item/types";

interface IMealItemProps {
  isDetails: boolean;
  item: Item;
  deleteMealItem(name: string): void;
  itemToEdit(name: string): void;
}

const MealItem: React.FC<IMealItemProps> = props => {
  const { itemName, amount, unit } = props.item;
  return (
    <li className="itemlist__item">
      <p className="item__name">{itemName}</p>
      <p className="item__amount">
        {amount} {unit}
      </p>
      {!props.isDetails && (
        <div className="item__update">
          <i
            className="fas fa-edit item__update-edit"
            onClick={() => {
              props.itemToEdit(itemName);
            }}
          />
          <i
            className="fas fa-trash item__update-delete"
            onClick={() => props.deleteMealItem(itemName)}
          />
        </div>
      )}
    </li>
  );
};

export default MealItem;
