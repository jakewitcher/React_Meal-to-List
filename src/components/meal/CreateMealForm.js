import React from 'react';

const MealForm = (props) => {
    const {
        mealNameChange,
        itemNameChange,
        amountChange,
        unitChange,
        addItem,
        resetMeal,
        mealName,
        itemName,
        amount,
        unit,
        itemList,
        dispatch
    } = props.mealProps;
    const { addMeal } = props;

    return (
        <div className="form">
            <h2 className="form__header">Create a New Meal</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Meal Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={mealName}
                        onChange={mealNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Item</p>
                        <input
                            className="form__iteminputfields--input"
                            type="text"
                            value={itemName}
                            onChange={itemNameChange}
                        />
                    </div>

                    <div>
                        <p>Amount</p>
                        <input
                            className="form__iteminputfields--input"
                            type="number"
                            value={amount}
                            onChange={amountChange}
                        />
                    </div>

                    <div>
                        <p>Unit</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="unit"
                            value={unit}
                            onChange={unitChange}
                        >
                            <option value="pound(s)">pound(s)</option>
                            <option value="ounce(s)">ounce(s)</option>
                            <option value="gallon(s)">gallons</option>
                            <option value="quart(s)">quart(s)</option>
                            <option value="pint(s)">pint(s)</option>
                            <option value="item(s)">item(s)</option>
                            <option value="bag(s)">bag(s)</option>
                            <option value="box(es)">box(es)</option>
                            <option value="container(s)">container(s)</option>
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addItem}
                        >
                            Add Item
                        </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            const name = mealName;
                            dispatch(addMeal({
                                name,
                                itemList
                            }));
                            resetMeal();
                        }}
                    >
                        Save Meal
                    </button>
                </div>
            </div>
        </div>
    )
}



export default MealForm;