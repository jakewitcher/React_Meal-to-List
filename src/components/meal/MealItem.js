import React from 'react';

const MealItem = () => {
    return (
        <li className="itemlist__item">
            <div className="item">
                <p>name</p>
            </div>
            <div className="item">
                <p>amount</p>
            </div>
            <div className="item">
                <p>unit</p>
            </div>
        </li>
    );
}

export default MealItem;