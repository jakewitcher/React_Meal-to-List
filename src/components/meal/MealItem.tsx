import React from "react";

interface IMealItemProps {
  isDetails: boolean;
  item: { itemName: string; amount: number; unit: string };
  deleteMealItem(name: string): void;
  itemToEdit(name: string): void;
}

const MealItem = (props: IMealItemProps) => {
  const { itemName, amount, unit } = props.item;
  return (
    <li className="itemlist__item">
      <p className="item__name">{itemName}</p>
      <p className="item__amount">
        {amount} {unit}
      </p>
      {!props.isDetails && (
        <div className="item__update">
          <i
            className="fas fa-edit item__update-edit"
            onClick={() => {
              props.itemToEdit(itemName);
            }}
          />
          <i
            className="fas fa-trash item__update-delete"
            onClick={() => props.deleteMealItem(itemName)}
          />
        </div>
      )}
    </li>
  );
};

export default MealItem;
