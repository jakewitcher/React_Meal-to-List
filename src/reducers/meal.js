import { defaultMealsList } from './seed';

// const mealDefaultState = {
//     mealList: [],
// };

export default (state = {mealList: defaultMealsList}, action) => {
    switch (action.type) {
        case 'ADD_MEAL':
            return {
                ...state,
                mealList: [...state.mealList, action.meal]
            };
        default:
            return state;
    }
}