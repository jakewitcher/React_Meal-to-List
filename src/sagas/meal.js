import { put, takeEvery, select } from 'redux-saga/effects';
import database from '../firebase/firebase';

const getUserId = (state) => state.auth.uid;

const formatItemList = (list) => {
    return list.reduce((list, item) => {
        list[item.itemName] = {
            amount: item.amount,
            unit: item.unit
        };
        return list;
    }, {});
};

function* addMeal({ meal = {} } = {}) {
    const { itemList, name } = meal;
    const items = formatItemList(itemList);

    let id;
    const uid = yield select(getUserId);
    yield database.ref(`users/${uid}/meals`).push({ name, items }).then((ref) => {
        id = ref.key;
    });

    yield put({
        type: 'ADD_MEAL',
        meal: {
            name,
            itemList,
            id,
        }
    });
}

function* editMeal({ meal = {} }) {
    const { itemList, name, id } = meal;
    const items = formatItemList(itemList);
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals/${id}`).update({ name, items });

    yield put({
        type: 'EDIT_MEAL',
        meal: {
            name,
            itemList,
            id,
        }
    });
};

function* deleteMeal({ id = '' }) {
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals/${id}`).remove();

    yield put({
        type: 'DELETE_MEAL',
        id,
    });
}

function* setMeal() {
    const meals = [];
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals`).once('value').then((snapshot) => {
        snapshot.forEach((childSnapshot) => {
            const itemList = [];
            childSnapshot.child('items').forEach((item) => {
                itemList.push({
                    itemName: item.key,
                    amount: item.val().amount,
                    unit: item.val().unit,
                });
            });

            meals.push({
                id: childSnapshot.key,
                name: childSnapshot.val().name,
                itemList,
            });
        });
    });

    yield put({ type: 'SET_MEAL', meals });
};

export function* watchAddMeal() {
    yield takeEvery('ON_ADD_MEAL', addMeal);
};

export function* watchEditMeal() {
    yield takeEvery('ON_EDIT_MEAL', editMeal);
};

export function* watchDeleteMeal() {
    yield takeEvery('ON_DELETE_MEAL', deleteMeal);
};

export function* watchSetMeal() {
    yield takeEvery('ON_SET_MEAL', setMeal);
}