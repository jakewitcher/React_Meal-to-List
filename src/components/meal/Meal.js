import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { onDeleteMeal } from '../../actions/meal';

export const Meal = (props) => {
    const { name, id, itemList } = props;
    return (
        <div className="list-type">
            <Link className="list-type__link" to={`/meals/meal/${id}`}>
                <p className="list-type__name">{name}</p>
            </Link>
            <p className="list-type__details">number of items: <span className="list-type__details list-type__details--span"> {itemList.length}</span></p>
            <div className="list-type__update">
                <Link className="list-type__link" to={`/meals/edit/${id}`}>
                    <i className="list-type__edit fas fa-edit" />
                </Link>
                <i
                    className="list-type__delete fas fa-trash"
                    onClick={() => props.onDeleteMeal(id)}
                />
            </div>
        </div>
    );
}

const mapDispatchToProps = (dispatch) => {
    return {
        onDeleteMeal: (id) => dispatch(onDeleteMeal(id)),
    }
}

export default connect(undefined, mapDispatchToProps)(Meal);