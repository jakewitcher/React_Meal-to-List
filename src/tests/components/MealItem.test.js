import React from 'react';
import { shallow } from 'enzyme';
import MealItem from '../../components/meal/MealItem';
import { mealItems } from '../fixtures/meal';

let item, deleteMealItem, itemToEdit, wrapper;

beforeEach(() => {
    item = {
        itemName: mealItems[0].itemName,
        amount: mealItems[0].amount,
        unit: mealItems[0].unit,
    };
    deleteMealItem = jest.fn();
    itemToEdit = jest.fn();
    wrapper = shallow(
        <MealItem 
            item={item}
            deleteMealItem={deleteMealItem}
            itemToEdit={itemToEdit}
        />
    );
});

test('should render MealItem correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should delete meal item by id', () => {
    wrapper.find('i').at(1).simulate('click');
    expect(deleteMealItem).toHaveBeenLastCalledWith(item.itemName);
});

test('should select meal item to edit by id', () => {
    wrapper.find('i').at(0).simulate('click');
    expect(itemToEdit).toHaveBeenLastCalledWith(item.itemName);
});