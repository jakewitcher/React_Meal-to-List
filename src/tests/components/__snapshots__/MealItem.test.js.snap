// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MealItem correctly 1`] = `
<li
  className="itemlist__item"
>
  <p
    className="item__name"
  >
    bacon
  </p>
  <p
    className="item__amount"
  >
    2
     
    pounds
  </p>
  <div
    className="item__update"
  >
    <i
      className="fas fa-edit item__update--edit"
      onClick={[Function]}
    />
    <i
      className="fas fa-trash item__update--delete"
      onClick={[Function]}
    />
  </div>
</li>
`;
